src/library/features/crates/cratehierarchy.cpp

```diff
@@ -0,0 +1,274 @@
+}
+
+void CrateHierarchy::resetClosure() const {
+    FwdSqlQuery query(
+      m_database, QString(
+        "DELETE FROM %1").arg(
+          CRATE_CLOSURE_TABLE));
+    if (!query.isPrepared()) {
+        return;
+    }
+    if (!query.execPrepared()) {
+        return;
+    }
+}
+
+
+bool CrateHierarchy::initClosure() const {
+    std::vector<CrateId> crateIds;
+
+    CrateSelectResult crates(m_pTrackCollection->crates().selectCrates());
+    Crate crate;
+
+    while (crates.populateNext(&crate)) {
+        crateIds.push_back(crate.getId());
+    }
+
+    FwdSqlQuery query(
+      m_database, QString(
+        "INSERT INTO %1 VALUES("
+        ":parent, :child, 0)").arg(
+          CRATE_CLOSURE_TABLE));
+    if (!query.isPrepared()) {
+        return false;
+    }
+
+    while (crateIds.size() != 0) {
+        query.bindValue(":parent", crateIds.back());
+        query.bindValue(":child", crateIds.back());
+        if (!query.execPrepared()) {
+            return false;
+        }
+        crateIds.pop_back();
+    }
+
+    return true;
+}
+
+void CrateHierarchy::resetPath() const {
+    FwdSqlQuery query(
+      m_database, QString(
+        "DELETE FROM %1").arg(
+          CRATE_PATH_TABLE));
+    if (!query.isPrepared()) {
+        return;
+    }
+    if (!query.execPrepared()) {
+        return;
+    }
+}
+
+bool CrateHierarchy::writeCratePaths(CrateId id, QString namePath, QString idPath) const {
+    FwdSqlQuery query(
+      m_database, QString(
+        "INSERT INTO %1 "
+        "VALUES (:id, :idPath, :namePath)").arg(
+          CRATE_PATH_TABLE));
+    if (!query.isPrepared()) {
+        return false;
+    }
+    query.bindValue(":id", id);
+    query.bindValue(":namePath", namePath);
+    query.bindValue(":idPath", idPath);
+    if (!query.execPrepared()) {
+        return false;
+    }
+
+    return true;
+}
+
+bool CrateHierarchy::generateCratePaths(Crate crate) const {
+    QSqlQuery query(m_database);
+
+    query.prepare(
+      QString(
+        "SELECT p.%1, p.%2 FROM %3 "
+        "JOIN %4 p ON %5 = p.%2 "
+        "JOIN %4 c ON %6 = c.%2 "
+        "where c.%1 = :childName and %7 != 0 "
+        "ORDER BY %7 DESC").arg(
+          CRATETABLE_NAME,
+          CRATETABLE_ID,
+          CRATE_CLOSURE_TABLE,
+          CRATE_TABLE,
+          CLOSURE_PARENTID,
+          CLOSURE_CHILDID,
+          CLOSURE_DEPTH));
+
+    query.bindValue(":childName", crate.getName());
+    query.setForwardOnly(true);
+    QString namePath;
+    QString idPath;
+
+    if (query.exec()) {
+        while (query.next()) {
+            namePath = namePath + "/" + query.value(0).toString();
+            idPath = idPath + "/" + query.value(1).toString();
+        }
+    } else {
+        return false;
+    }
+
+    namePath = namePath + "/" + crate.getName();
+    idPath = idPath + "/" + crate.getId().toString();
+
+    if(!writeCratePaths(crate.getId(), namePath, idPath)) {
```

Comment:
This can be simplified to `return writeCratePaths(crate.getId(), namePath, idPath);`